crypto/fipsmodule/evp/digestsign.c

```diff
@@ -164,22 +177,34 @@ int EVP_DigestSignFinal(EVP_MD_CTX *ctx, uint8_t *out_sig,
     FIPS_service_indicator_lock_state();
 
     EVP_MD_CTX_init(&tmp_ctx);
-    ret = EVP_MD_CTX_copy_ex(&tmp_ctx, ctx) &&
-          EVP_DigestFinal_ex(&tmp_ctx, md, &mdlen) &&
-          EVP_PKEY_sign(ctx->pctx, out_sig, out_sig_len, md, mdlen);
+    if (used_for_hmac(ctx)) {
+      // This is only defined in |EVP_PKEY_HMAC|.
+      ret = EVP_MD_CTX_copy_ex(&tmp_ctx, ctx) &&
+            tmp_ctx.pctx->pmeth->hmac_final(tmp_ctx.pctx, out_sig, out_sig_len,
+                                           &tmp_ctx);
+    } else {
+      ret = EVP_MD_CTX_copy_ex(&tmp_ctx, ctx) &&
+            EVP_DigestFinal_ex(&tmp_ctx, md, &mdlen) &&
+            EVP_PKEY_sign(ctx->pctx, out_sig, out_sig_len, md, mdlen);
+    }
     EVP_MD_CTX_cleanup(&tmp_ctx);
 
     FIPS_service_indicator_unlock_state();
-    if(ret > 0) {
+    if (ret > 0) {
       EVP_DigestSign_verify_service_indicator(ctx);
     }
     return ret;
   } else {
     // This only determines the size of the signature. This case of
     // |EVP_DigestSignFinal| should not return an approval check because no
     // crypto is being done.
-    size_t s = EVP_MD_size(ctx->digest);
-    return EVP_PKEY_sign(ctx->pctx, out_sig, out_sig_len, NULL, s);
+    if (used_for_hmac(ctx)) {
+      // This is only defined in |EVP_PKEY_HMAC|.
+      return ctx->pctx->pmeth->hmac_final(ctx->pctx, out_sig, out_sig_len, ctx);
```

Comment:
could really just be `return EVP_MD_CTX_size(ctx)`. Or a bit more descriptive `return HMAC_size(EVP_MD_CTX_get_hmac(ctx))`, if wanting to add more functions.